cdc/log.cc

```diff
@@ -932,8 +932,12 @@ class log_mutation_builder {
     // Given a reference to such a column from the base schema, this function sets the corresponding column
     // in the log to the given value for the given row.
     void set_value(const clustering_key& log_ck, const column_definition& base_cdef, const managed_bytes_view& value) {
-        auto& log_cdef = *_log_schema.get_column_definition(log_data_column_name_bytes(base_cdef.name()));
-        _log_mut.set_cell(log_ck, log_cdef, atomic_cell::make_live(*base_cdef.type, _ts, value, _ttl));
+        auto log_cdef = _log_schema.get_column_definition(log_data_column_name_bytes(base_cdef.name()));
+        // If table uses compact storage it may contain a column of type empty
+        // and we need to ignore such a field because it is not present in CDC log.
+        if (log_cdef) {
+            _log_mut.set_cell(log_ck, *log_cdef, atomic_cell::make_live(*base_cdef.type, _ts, value, _ttl));
+        }
```

Comment:
For deltas you earlier fixed the problem by skipping the column at call site, in `process_row_visitor::live_atomic_cell`:

```cpp
    void live_atomic_cell(const column_definition& cdef, const atomic_cell_view& cell) {
        _ttl_column = get_ttl(cell);

        if (cdef.type->get_kind() == abstract_type::kind::empty) {
            return;
        }

        managed_bytes value = get_managed_bytes(cell);

        // delta
        if (_generate_delta_values) {
            _builder.set_value(_log_ck, cdef, value);
        }
```

For consistency, I think preimages should be solved the same way, i.e. by skipping the column at call site, in this case `transformer::generate_images`:

```cpp
        auto process_cell = [&, this] (const column_definition& cdef) {

---> here we could skip empty type columns

            if (auto current = get_col_from_row_state(row_state, cdef)) {
                _builder->set_value(image_ck, cdef, *current);
            } else if (op == operation::pre_image) {
                // Cell is NULL. 
                // If we generate the preimage,
                // we should also fill in the deleted column.
                // Otherwise the user will not be able
                // to discern whether a value in the preimage
                // is NULL or it was not included in the
                // preimage ('full' preimage disabled).
                // If we generate the postimage,
                // we don't have to fill in the deleted column,
                // as all postimages are full.
                _builder->set_deleted(image_ck, cdef);
            }
        };

        if (affected_columns) {
            // Preimage case - include only data from requested columns
            for (auto it = affected_columns->find_first(); it != one_kind_column_set::npos; it = affected_columns->find_next(it)) {
                const auto id = column_id(it);
                const auto& cdef = _schema->column_at(kind, id);
                process_cell(cdef);
            }
        } else {
            // Postimage case - include data from all columns
            const auto column_range = _schema->columns(kind);
            std::for_each(column_range.begin(), column_range.end(), process_cell);
        }
```

besides I'm worried if the `else` branch inside `process_cell` couldn't happen for this compact column, which would cause `set_deleted` to be called, which would set a value in the CDC log mutation for a column that doesn't exist...

Btw, the fact that we tried to generate preimage value for this column means that `row_state` had a value for this column. Which means we had to set it in `load_preimage_results_into_state`. Perhaps we should also skip empty type columns there as well.

Reply:
Ok. I changed the place of the check.

Reply:
But I don't see how `load_preimage_results_into_state` is relevant to `generate_image`.

Reply:
`process_cell` for preimage is called only on `affected_columns` which are constructed in `process_changes_with(out)_splitting`.

Reply:
>  But I don't see how load_preimage_results_into_state is relevant to generate_image.

`load_preimage_results_into_state` generates the initial values of `_static_row_state` and `_clustering_row_states`, which are then used to generate preimages and updated as we iterate through the changes.

The fact that we called `_builder->set_value(...)` means that this `if` passed:
```
            if (auto current = get_col_from_row_state(row_state, cdef)) {
```
i.e. our state (it was `_clustering_row_state` in this case I guess) had a value for this empty type column. Which means that either `load_preimage_results_into_state` loaded a value there, or it was updated when visiting base mutations cells, but the second case couldn't happen since we skip empty type columns in `live_atomic_cell`.

> process_cell for preimage is called only on affected_columns which are constructed in process_changes_with(out)_splitting.

`affected_columns` is also `nullptr` for full preimages and for postimages, in that case `process_cell` is called for every column.